feat(agent): add agent socket API

SasSwart opened the pull request:
relates to: https://github.com/coder/internal/issues/1094

This is number 2 of 5 pull requests in an effort to add agent script ordering. It adds a drpc API that is exposed via a local socket. This API serves access to a lightweight DAG based dependency manager that was inspired by systemd.

In follow-up PRs:

* This unit manager will be plumbed into the workspace agent struct.
* CLI commands will use this agentsocket api to express dependencies between coder scripts

I used an LLM to produce some of these changes, but I have conducted thorough self review and consider this contribution to be ready for an external reviewer.

SasSwart committed:
feat(agent): add agent socket server infrastructure

SasSwart committed:
appease the linter

SasSwart committed:
Fix build issues post rebase

SasSwart committed:
dont regenerate the agentsocket files when we build dylib

SasSwart committed:
fix unit manager service tests

SasSwart committed:
Use mac compatible socket paths for agentsocket service testing

SasSwart committed:
improve macos socket path helper

SasSwart committed:
fix function signatures for the agentsocket api on windows

SasSwart committed:
disable the agentsocket API on windows

SasSwart committed:
Disable agentsocket api on windows

SasSwart committed:
adjust to updates from upstream after rebase

SasSwart committed:
add and mature agentsocket tests

mafredri commented on agent/agentsocket/proto/agentsocket.proto:
```diff
@@ -0,0 +1,89 @@
+  // Request whether a unit is ready to be started. That is, all dependencies are satisfied.
+	rpc SyncReady(SyncReadyRequest) returns (SyncReadyResponse);
+  // Get the status of a unit and list its dependencies.
+	rpc SyncStatus(SyncStatusRequest) returns (SyncStatusResponse);
```
Is the flow going to be:

- SyncWant(fooscript, barscript)
- SyncWant(fooscript, bazscript)
- while ! SyncReady(fooscript)
- SyncStart(fooscript)
- ...
- SyncComplete(fooscript)

I assume SyncStatus is more useful or debugging than actual script writing (which is fine).

mafredri replied:
One more thought, what happens if I reverse the order of want and start? So start -> want want -> ready. (I imagine users will feel it's natural to start immediately once your script gets going, it may be wrong use but still.)

SasSwart replied:
Your described flow is correct. Although SyncStart should probably do the polling server side.

In terms of watch vs poll, we agreed with Cian to only poll for now. We can implement watch later.

To your last thought:
* Want should almost always be called before Start.
* Wants that are added after Start will only be applied on the next Start.
* We could add a guard to say that dependencies cannot be added to a unit that has been started.

mafredri commented on agent/agentsocket/server.go:
```diff
@@ -0,0 +1,185 @@
+		s.path, err = getDefaultSocketPath()
+		if err != nil {
+			return xerrors.Errorf("get default socket path: %w", err)
+		}
```
New already returns an error, we should rather error early if we can.

I'd say the same for context, no need to defer its assignment to here.

If the purpose is to support Start/Stop/Start, then `s.path` assignment is destructive breaks Stops ability to be a clean reset.

I'd recommend just creating a new manager if it needs to be reset but there may be a reason not to?

mafredri commented on agent/agentsocket/service.go:
```diff
@@ -0,0 +1,260 @@
+	if err := s.unitManager.AddDependency(req.Unit, req.DependsOn, unit.StatusComplete); err != nil {
+		return &proto.SyncWantResponse{
+			Success: false,
+			Message: "failed to add dependency: " + err.Error(),
```
Nit: Mixed use of upper/lower case, please standardize.

mafredri commented on agent/agentsocket/socket_unix.go:
```diff
@@ -0,0 +1,76 @@
+
+	// Fall back to /tmp with user-specific path
+	uid := os.Getuid()
+	return filepath.Join("/tmp", fmt.Sprintf("coder-agent-%d.sock", uid)), nil
```
(I wonder if this might create conflicts between tests. Especially later on with CLI support and CLI tests. Consider adding a CLI flag to configure it.)

mafredri commented on agent/agentsocket/socket_windows.go:
```diff
@@ -0,0 +1,53 @@
+// createSocket returns an error indicating that agentsocket is not supported on Windows.
+// This feature is unix-only in its current experimental state.
+func createSocket(path string) (net.Listener, error) {
+	return nil, xerrors.New("agentsocket is not supported on Windows")
```
IIRC supporting Windows should be possible, at least to some degree. We have other socket support in the product for Windows but can't recall if it has limitations.

SasSwart committed:
reduce overhead on the agentsocket api

SasSwart committed:
adjust the agentsocket API to the unit manager updates

SasSwart commented on agent/agentsocket/service.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	if err := s.unitManager.Register(req.Unit); err != nil {
+		// If already registered, that's okay - we can still update status
```
We update the status further down. I take your point though. The comment was vague and unnecessary. Removed.

SasSwart commented on agent/agentsocket/proto/agentsocket.proto:
```diff
@@ -0,0 +1,89 @@
+
+message SyncStartResponse {
+	bool success = 1;
+	string message = 2;
```
I've trimmed down this API. Please check again?

SasSwart committed:
Move unit validation into the unit manager

SasSwart committed:
Further slim down the agentsocket proto

DanielleMaywood commented on agent/agentsocket/server.go:
```diff
@@ -0,0 +1,185 @@
+func (s *Server) handleConnection(conn net.Conn) {
+	defer conn.Close()
+
+	if err := conn.SetDeadline(time.Now().Add(30 * time.Second)); err != nil {
```
+1 on reason for magic 30 seconds

DanielleMaywood commented on agent/agentsocket/service_test.go:
```diff
@@ -0,0 +1,503 @@
+	cleanup := func() {
+		_ = session.Close()
+		_ = conn.Close()
+	}
```
Why not just register `t.Cleanup` here instead of requiring a `defer` or `t.Cleanup` at the consumer level?

SasSwart committed:
remove defunct code and protect against a race

SasSwart committed:
Review notes

Further slim down the agentsocket proxy

simplify the agentsocket server interface

avoid two race conditions

use t.Cleanup where applicable

SasSwart committed:
make fmt

mafredri commented on agent/agentsocket/socket_unix.go:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	// Try to connect to see if it's actually listening
+	conn, err := net.Dial("unix", path)
```
Can we use `net.Dialer{}.DialContext(ctx, ...)` here so we don't block forever if the socket is in a bad state or there's a bad agent listening on it?

SasSwart committed:
Add a test to ensure that duplicate dependencies are not added

SasSwart committed:
consistent indentation for agentsocket.proto

SasSwart committed:
Add a timeout to socket detection

make gen

SasSwart committed:
Default to a random agentsocket path to avoid conflicts in tests. Production agents must specify a well-known agentsocket path and clients must be able to specify the socket to address

DanielleMaywood reviewed (approved):
stamping with a nit about a comment

SasSwart committed:
Rename outdated comment

mtojek reviewed (commented):
post-merge LGTM! Thank you both
